app/fork_manager.go

```diff
@@ -0,0 +1,65 @@
+package app
+
+import (
+	sdk "github.com/cosmos/cosmos-sdk/types"
+)
+
+type HardForkHandler interface {
+	GetTargetChainID() string
+	GetTargetHeight() int64
+	ExecuteHandler(ctx sdk.Context) error
+}
+
+type HardForkManager struct {
+	chainID    string
+	handlerMap map[int64][]HardForkHandler
+}
+
+// Create a new hard fork manager for a given chain ID.
+// This will filter out any handlers for other chain IDs,
+// and create a map that maps between target height and handlers to run for the given heights.
+func NewHardForkManager(chainID string) *HardForkManager {
+	return &HardForkManager{
+		chainID:    chainID,
+		handlerMap: make(map[int64][]HardForkHandler),
+	}
+}
+
+// This tries to register a handler with the hard fork manager.
+// If the handler's target chain ID doesn't match the chain ID of the manager,
+// this is a no-op and the handler is ignored.
+func (hfm *HardForkManager) RegisterHandler(handler HardForkHandler) {
+	if handler.GetTargetChainID() != hfm.chainID {
+		return
+	}
+	targetHeight := handler.GetTargetHeight()
+	if handlers, ok := hfm.handlerMap[targetHeight]; ok {
+		// TODO: refactor to use goutils InPlaceAppend once it's merged
+		handlers = append(handlers, handler)
```

Comment:
I wonder if we should enforce handlers to have a name `Name() string` as to ensure we don't register duplicate handlers or something?

Reply:
yeah, I think this is reasonable, should be pretty easy to add in. would we want the name to be unique regardless of target height though? I would think yes because the premise would be to have a name correspond to a specific remediation. Therefore if you want to do a different remediation at some different height, you would want the name to still be unique. I'll add a `Name` function to the interface